onnxruntime/core/session/inference_session.cc

```diff
@@ -1984,13 +1984,15 @@ static void ResolveMemoryPatternFlags(SessionState& session_state) {
 // For now, this function only checks for invalid combination of DML EP with other EPs.
 // TODO: extend this function to check for other invalid combinations of EPs.
 common::Status InferenceSession::HasInvalidCombinationOfExecutionProviders() const {
-  // DML EP is only allowed with CPU EP
+  // DML EP is not allowed with other GPU or NPU EPs.
+  // historical reason for this is unknown. relaxing the limit that it must only be used with the CPU EP to support
```

Comment:
```suggestion
  // The historical reason for this is unknown. relaxing the limit that it must only be used with the CPU EP to support
```
Yeah, I forget. Pranav might recall, but alas he left MS. I think part of the concern was that the combo had never been tested before (since the ORT with WinML only ever used DML), and that ORT might not do GPU-to-GPU copies between partitions nicely. I wonder how robustly ORT works with *other* heteregenous combinations, like say a graph containing nodes partitioned to OpenVINO and CUDA (or ROCM or TensorRT)?

Reply:
The memcpy transformer handles copying between a non-CPU OrtDevice and CPU. In general that works perfectly fine as there's usually (at least previously) only the CPU EP and an EP for a non-CPU device.

Compiling EPs tend to register with CPU based memory (or host accessible) so there's no device copy required by ORT there. 

In theory you could do something like enable an EP with static kernels for a discrete GPU (let's say CUDA EP for an nvidia card) and a different EP with static kernels for an integrated GPU (let's say DML). if you did that, and the model got partitioned and data needed to go from the dGPU to the iGPU for two adjacent nodes it wouldn't work as there's no special-casing to determine it needs to do that via CPU memory. 

but given that doesn't feel like a realistic scenario, adding that special-casing feels like it's a lot of cost for possibly zero benefit.